Add the 'Virtual Environments in Data' blog post (#688)
TobikoData/sqlmesh: posts/virtual_environments.md

# Virtual Environments in Data
*by [Iaroslav Zeigerman](https://github.com/izeigerman)*

We start this post by setting context for what environments are and how they are used in software today. If you want to skip this part and jump right in please proceed to the [Environments in Data](#environments-in-data) section.

## Environments in Software

> Environments have always been a way for software engineers to test and promote their changes in a safe manner, reducing this way the possibility of causing regressions in production. Generally each environment represents a set of isolated resources to which the software in question is deployed.

Reviewer: ```suggestion
Environments have always been a way for software engineers to test and deploy their changes safely, reducing the possibility of causing regressions in production. Generally each environment represents a set of isolated resources to which the software in question is deployed.
```

Over time the industry settled on a certain set of environments that is now considered to be the best practice:

> - **Testing** - used for testing the bleeding edge of the target software. This environment is updated frequently with most recent code changes. Deployments to this environment are low risk since it's only used by developers of a specific team and is never exposed to end users.

Reviewer: ```suggestion
- **Development** - used for testing the bleeding edge of the target software. This environment is updated frequently with most recent code changes. Deployments to this environment are low risk since it's only used by developers of a specific team and is never exposed to end users.
```

> - **Staging** - used for testing changes that are about to be released and promoted to production. Deployment to this environment happens shortly before the release and is usually only updated with critical bug fixes until the release takes place. Deployments to this environment are of a higher risk since it can be shared across multiple teams and/or organizations and can even be sometimes exposed to a limited set of partners and/or consumers.

Reviewer: ```suggestion
- **Staging** - used for testing changes that are about to be released and deployed to production. Deployment to this environment happens shortly before the release and is usually only updated with critical bug fixes until the release takes place. Deployments to this environment are of a higher risk since it can be shared across multiple teams and/or organizations and can even be sometimes exposed to a limited set of partners and/or consumers.
```

- **Production** - the name speaks for itself. Only released changes are promoted to this environment, usually after going through extensive testing and validation on staging. Deployments to this environment are of highest risk since it's the one exposed to consumers.

These environments, however, used to be quite rigid. The process of provisioning new environments or updating existing ones used to take a lot of time and required a fair bit of coordination across teams or even within a single team. Long story short, the process hardly scaled.

This changed with the rise of Docker and CI/CD which made the process of spinning up new testing environments extremely cheap and easy. This, in turn, resulted in a significant boost of productivity for individual developers, who could now easily provision new testing environments specifically for their needs and then tear them down as easily once they no longer needed them. Similarly, promoting tested changes to staging and production became equally straightforward.

## Environments in Data

> Environments have been routinely used by practitioners from the data analysis and data engineering fields. Tools like **dbt** allow its users to configure different *targets* - connections to different schemas within a Data Warehouse or even separate Data Warehouses instances altogether. This mechanism is sufficient to enable analysts and engineers to deploy their data pipelines into separate environments for testing, staging and production, while ensuring that changes are only applied to the target environment keeping tables in other environments unaffected. All such environments are usually provisioned once and then shared between all users who use the tool.

Reviewer: a bit confused about this. in dbt, you create your own dev environment which is usually a full copy right, what do you mean shared between all users who use the tool
